File: node/scheduler.js
/**
 * Polls for due reminders and fires them. Kept separate from bot.js
 * so the polling interval itself is the only untestable part --
 * everything it calls (dueReminders, markSent) is pure and already
 * tested in reminders-logic.test.js.
 */

import { formatReminderFired } from "./discord-format.js";
import { dueReminders, markSent } from "./reminders-logic.js";
import { markReminderSent } from "./reminders-store.js";

export function startReminderScheduler(client, ctx, intervalMs = 15_000) {
  return setInterval(async () => {
    const due = dueReminders(ctx.reminders);
    for (const reminder of due) {
      try {
        const channel = await client.channels.fetch(reminder.channelId);
        await channel.send(formatReminderFired(reminder));
      } catch (err) {
        console.error(`Impossible d'envoyer le rappel ${reminder.id} :`, err.message);
      }
      markReminderSent(ctx.reminderDb, reminder.id);
      ctx.reminders = markSent(ctx.reminders, reminder.id);
    }
  }, intervalMs);
}

File: python/scheduler.py
"""
Polls for due reminders and fires them. Kept separate from bot.py so
the polling interval itself is the only untestable part -- everything
it calls (due_reminders, mark_sent) is pure and already tested in
test_reminders_logic.py.
"""

import asyncio
import logging

from reminders_logic import due_reminders, mark_sent
from reminders_store import mark_reminder_sent
from telegram_format import format_reminder_fired

logger = logging.getLogger(__name__)


async def run_reminder_scheduler(bot, ctx, interval_seconds=15):
    while True:
        await asyncio.sleep(interval_seconds)
        due = due_reminders(ctx["reminders"])
        for reminder in due:
            try:
                await bot.send_message(chat_id=int(reminder["channel_id"]), text=format_reminder_fired(reminder))
            except Exception as err:  # best-effort delivery -- log and keep going
                logger.error("Impossible d'envoyer le rappel %s : %s", reminder["id"], err)
            mark_reminder_sent(ctx["reminder_db"], reminder["id"])
            ctx["reminders"] = mark_sent(ctx["reminders"], reminder["id"])
